include/mitsuba/render/optix/shapes.h

```diff
@@ -27,5 +32,213 @@ size_t get_shape_descr_idx(Shape *shape) {
     Throw("Unexpected shape: %s. Couldn't be found in the "
           "'custom_optix_shapes' table.", name);
 }
+
+struct OptixAccelData {
+    OptixTraversableHandle handle = 0ull;
+    void* buffer_meshes = nullptr;
+    void* buffer_others = nullptr;
+    void* buffer_ias = nullptr;
+
+    ~OptixAccelData() {
+        if (handle) {
+            cuda_free(buffer_meshes);
+            cuda_free(buffer_others);
+            cuda_free(buffer_ias);
+        }
+    }
+};
+
+template <typename Shape>
+void fill_hitgroup_records(std::vector<HitGroupSbtRecord> &hitgroup_records,
+                           std::vector<ref<Shape>> &shapes,
+                           OptixProgramGroup *program_groups) {
+    for (size_t i = 0; i < 2; i++)
+        for (Shape* shape: shapes)
+            if (i == !shape->is_mesh())
+                shape->optix_fill_hitgroup_records(hitgroup_records, program_groups);
```

Comment:
Let's add the braces expect for the `if` statement here. Also maybe add a comment that meshes a processed first.